### ISSUE TITLE
Add rewrite frames integration

### ISSUE THREAD

**Issue by denrase** [getsentry/sentry-dart#2111]
We could add a rewrite frames integration similar to [javascript](https://docs.sentry.io/platforms/javascript/configuration/integrations/rewriteframes/) that allows users to manipulate the frames in a specific way before sending them off.

### Examples of what it would solve

Relates to https://github.com/getsentry/sentry-dart-plugin/pull/222

When users provide a `urlPrefix` with `sentry-dart-plugin` and `sentry-cli`, stack traces are not resolved correctly, as the path cant be found.

the stack trace abs path is `"http://localhost: 8900/main.dart`.
the uploaded file is `~/foo/main.dart` -> doesn't match

so you would need to rewrite the abs_path to the uploaded file to include the prefix.

With a rewrite frame integration it would be easy to do that.

### Alternatives

Use beforeSend code snippet

```dart
beforeSendCallback: (event, hint) async {
  final exceptions = event.exceptions?.map((exception) {
    final stackTrace = exception.stackTrace;
    if (stackTrace != null) {
      final frames = stackTrace.frames.map((frame) {
        const baseUrl = 'https://example.com/';
        final modifiedAbsPath = frame.absPath?.replaceFirst(baseUrl, '${baseUrl}my_prefix/');
        return frame.copyWith(absPath: modifiedAbsPath);
      }).toList();
      return exception.copyWith(stackTrace: SentryStackTrace(frames: frames));
    }
    return exception;
  }).toList() ?? [];
  return event.copyWith(exceptions: exceptions);
});
```

### Demand

Please upvote if you are interested in such an integration

**Comment by denrase** [comment]
@buenaflor I created this issue as we only talked in Discord about this. Is this still correct and something we want to do?

**Comment by buenaflor** [comment]
Yep, the most simplest way is to tell users to modify the abs path of the stackframes in `beforeSend` to add the prefix themselves.  or we can add an integration like javascript for rewriting frames which is syntactically nicer: https://docs.sentry.io/platforms/javascript/configuration/integrations/rewriteframes/

I think we're fine in just adding the guide to configure it like so :

```dart
beforeSendCallback: (event, hint) async {
  final exceptions = event.exceptions?.map((exception) {
    final stackTrace = exception.stackTrace;
    if (stackTrace != null) {
      final frames = stackTrace.frames.map((frame) {
        const baseUrl = 'https://example.com/';
        final modifiedAbsPath = frame.absPath?.replaceFirst(baseUrl, '${baseUrl}my_prefix/');
        return frame.copyWith(absPath: modifiedAbsPath);
      }).toList();
      return exception.copyWith(stackTrace: SentryStackTrace(frames: frames));
    }
    return exception;
  }).toList() ?? [];
  return event.copyWith(exceptions: exceptions);
});
```

a specific rewrite frames integration is overkill for now

**Comment by denrase** [comment]
Added to docs: https://github.com/getsentry/sentry-docs/pull/10470

We could also add the sample in the sentry-dart-plugin

**Comment by buenaflor** [comment]
> We could also add the sample in the sentry-dart-plugin

let's do that

**Comment by dillon-giacoppo** [comment]
Since debug ID symbolication became the default in `sentry_dart_plugin` 3.0.0 / `sentry_flutter` 9.1.0, I don't think the documented `beforeSend` / [Rewrite Frames](https://docs.sentry.io/platforms/dart/guides/flutter/configuration/rewriteframes/) workaround works for this case (it's also documented in the dart-plugin [troubleshooting](https://docs.sentry.io/platforms/dart/guides/flutter/debug-symbols/dart-plugin/#troubleshooting)). The debug image is attached by an event processor that runs before `beforeSend`.

On web, [`SentryWeb.loadDebugImages`](https://github.com/getsentry/sentry-dart/blob/427c977e77dd64bd3cd66f716f22a48e3596eb64/packages/flutter/lib/src/web/sentry_web.dart#L176-L207) attaches the `DebugImage` by reading the injected `filename: debugId` map, then doing an [exact lookup](https://github.com/getsentry/sentry-dart/blob/427c977e77dd64bd3cd66f716f22a48e3596eb64/packages/flutter/lib/src/web/sentry_web.dart#L183-L187) against the first stack frame’s `absPath` / `fileName`. The map keys include the full asset URL, including the base path.

In a subpath deployment, the Dart SDK reports frames like:

```text
https://app.example.com/main.dart.js
```

but the injected debug-ID map contains:

```text
https://app.example.com/subpath/main.dart.js
```

Because the lookup is exact, no `DebugImage` is attached and the stack trace remains minified.

`loadDebugImages` runs via `_LoadDebugIdEventProcessor`, added by `LoadWebDebugImagesIntegration`. Since `SentryClient` runs event processors before `beforeSend`, rewriting `absPath` in `beforeSend` is too late. The rewrite needs to happen in an `EventProcessor` registered during init, before the integration processor runs.

**Workaround**

Register an event processor that rewrites same-origin frame `absPath`s using `document.baseURI`, including both exception and thread stack traces:

```dart
class BaseHrefStackFrameRewrite implements EventProcessor {
  BaseHrefStackFrameRewrite() : _base = Uri.parse(web.document.baseURI);

  final Uri _base;

  @override
  SentryEvent apply(SentryEvent event, Hint hint) {
    final basePath = _base.path.endsWith('/')
        ? _base.path.substring(0, _base.path.length - 1)
        : _base.path;

    if (basePath.isEmpty) return event;

    for (final exception in event.exceptions ?? const <SentryException>[]) {
      _rewrite(exception.stackTrace, basePath);
    }

    for (final thread in event.threads ?? const <SentryThread>[]) {
      _rewrite(thread.stacktrace, basePath);
    }

    return event;
  }

  void _rewrite(SentryStackTrace? stackTrace, String basePath) {
    if (stackTrace == null) return;

    for (final frame in stackTrace.frames) {
      final absPath = frame.absPath;
      if (absPath == null) continue;

      final uri = Uri.tryParse(absPath);
      if (uri == null ||
          uri.scheme != _base.scheme ||
          uri.host != _base.host ||
          uri.port != _base.port) {
        continue;
      }

      if (uri.path == basePath || uri.path.startsWith('$basePath/')) {
        continue;
      }

      frame.absPath = uri.replace(path: '$basePath${uri.path}').toString();
    }
  }
}

// options.addEventProcessor(BaseHrefStackFrameRewrite());
```

**Comment by buenaflor** [comment]
hey thanks for the details. yeah indeed the `beforeSend` snippet won't work anymore in that case.

We'll have a closer look and may incorporate your snippet in the docs instead
